api/mapping/forms.py

```diff
@@ -34,35 +35,65 @@ class ScanReportForm(forms.Form):
     )
     def clean_scan_report_file(self):
         if str(self.cleaned_data['scan_report_file']).endswith('.xlsx'):
-            xlsx = Xlsx2csv(self.cleaned_data['scan_report_file'], outputencoding="utf-8")
-
-            filepath = "/tmp/{}.csv".format(xlsx.workbook.sheets[0]["name"])
-            xlsx.convert(filepath)
-
-            with open(filepath, "rt") as f:
-                reader = csv.reader(f)
-                csv_header=next(reader)  # Get header row
-                set_header=['Table', 'Field', 'Description', 'Type', 'Max length', 'N rows', 'N rows checked', 'Fraction empty', 'N unique values', 'Fraction unique', 'Flag', 'Classification']
-                if set(set_header)==set(csv_header):
-                    for row in reader:
-                        flag_column=row[10]
-                        flag_column=flag_column.upper()
-                        classification_column=row[11]
-                        print(flag_column)
-                        if (flag_column in FLAG_CHOICES) or (flag_column==''):
-                            pass
-                        else:
-                            raise (forms.ValidationError( "Check Flag column values. Valid options are: {} or blank".format(list(FLAG_CHOICES.values()))))
-                        
-                        if (classification_column in VOCABULARY_CHOICES.values()) or (classification_column==''):
-                            pass
-                        else:
-                            raise (forms.ValidationError( "Check Classification column values. Valid options are:{} or blank".format(list(VOCABULARY_CHOICES.values()))))
-                    return self.cleaned_data['scan_report_file']
+
+            # Load in the Excel sheet, grab the first workbook
+            file_in_memory = self.cleaned_data['scan_report_file'].read()
+            wb = openpyxl.load_workbook(filename=BytesIO(file_in_memory), data_only = True)
+            ws=wb.worksheets[0]
+
+            # Grab the scan report columns from the first worksheet
+            # Define what the column headings should be
+            source_headers = []
+            for values in ws[1]: source_headers.append(values.value)
+            expected_headers=['Table', 'Field', 'Description', 'Type', 'Max length', 'N rows', 'N rows checked', 'Fraction empty', 'N unique values', 'Fraction unique', 'Flag', 'Classification']
+            
+            # Check if source headers match the expected headers
+            if set(source_headers) == set(expected_headers):
```

Comment:
`set()` doesn't preserve the ordering, so please change this to be a direct comparison of the lists - otherwise we risk ingesting files with all the correct columns but in the wrong order.

Reply:
And see the `if not` comment as above - drag the related `ValidationError` up to here.

Reply:
"as above" in this instance is in a comment further down here.